### PR TITLE
Changed default casefolding method to str.casefold; User-defined casefolding

### PR DESCRIPTION
**Pull request by andy-maier**
For details, see the commit message.

### CODE REVIEW

File: docs/intro.rst [hunk 1]
```diff
@@ -41,6 +41,11 @@ the following exceptions (and the case-insensitivity of course):
 
 .. _dict class of Python 3.8: https://docs.python.org/3.8/library/stdtypes.html#dict
 
+The case-insensitivity is achieved by matching any key values as their
+casefolded values. The casefolding is performed with :meth:`py:str.casefold` on
```
Comment (KSchopmeyer) [inline docs/intro.rst:45]: should it be default casefolding

File: docs/intro.rst [hunk 2]
```diff
@@ -53,6 +58,50 @@ Why yet another case-insensitive dictionary: We found that all previously
 existing case-insensitive dictionary packages on Pypi either had flaws, were
 not well maintained, or did not support the Python versions we needed.
 
+
+.. _`Overriding the casefold method`:
+
+Overriding the casefold method
```
Comment (KSchopmeyer) [inline docs/intro.rst:64]: should it be Overriding the default casefold method

File: docs/changes.rst [hunk 3]
```diff
@@ -20,6 +20,10 @@ Released: not yet
 
 * Added support for Python 3.11.
 
+* Changed the default casefolding method on Python 3 to be `str.casefold()`;
```
Comment (KSchopmeyer) [inline docs/changes.rst:23]: add "to improve unicode support"

**Review by KSchopmeyer — CHANGES_REQUESTED** [review]
In commit message and changes.rst (Made comment at that point in code) it would be helpful if we stated the reason for the change  (i.e. improved.

Question in general.  The cost of creating the class is greater in python 3 than python 2 because it builds several methods and then if not PY2, deletes them.  Could the opposite be to only conditionally create those methods if it is PY2? This would be logical since most usage in the future will be PY3. This is strictly a question of cost of setting up the class and I am not even sure it costs that much yet.

Should we state something in incompatibilities because of the move to casefold since there could be different results with the more aggressive selection of lower case characters?

File: docs/intro.rst (hunk 1 above)
Comment (andy-maier) [reply docs/intro.rst:45]: yes, DONE

File: docs/intro.rst (hunk 2 above)
Comment (andy-maier) [reply docs/intro.rst:64]: Yes, DONE

File: docs/changes.rst (hunk 3 above)
Comment (andy-maier) [reply docs/changes.rst:23]: DONE

**Comment by andy-maier** [comment]
On your other comments:

On 1: DONE in commit message and change log.

On 2: NO CHANGE. I did not find a way to add a method in such a way that Sphinx creates proper documentation for it. The extra cost is only on import of the nocasedict package, but not on creation of a NocaseDict object. Finally, the Python-2-only methods will go away once we implement type hints (issue #123).

On 3: DONE. I added an entry to the Incompatibilities section.

On 4: NO CHANGE. The nocasedict project also has that glossary entry.

**Comment by andy-maier** [comment]
Resolved Karl's comments.

**Comment by KSchopmeyer** [comment]
Approved  pending test completion

**Comment by andy-maier** [comment]
More changes:
* Added a fix by removing the incorrect statement that the input key of `__casefold__()`  can also be `None`.
* Applied the changes from intro.rst also to README.rst

**Comment by andy-maier** [comment]
I'll leave this PR open for a few more days to give @rudolfbyker a chance to review.

**Pull request merged**